tithemi/assignment2/Q2.cpp

```diff
@@ -0,0 +1,68 @@
+#include <array>
+#include <vector>
+#include <iostream>
+
+template<typename T, size_t N = 2>
+struct Node
+{
+    T value;
+    std::array<Node<T, N>*, N> children;
+
+    Node(const T value, std::array<Node<T, N>*, N> children)
+            : value(value), children(children)
+    { }
+
+    bool find(const T& target, std::vector<Node<T, N>*>& path)
+    {
+        if (value == target)
+            return true;
+
+        for (Node<T, N>* child : children)
+        {
+            if (child != nullptr && child->find(target, path))
+            {
+                path.push_back(this);
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    Node<T, N>* findCommon(const T& first, const T& second)
+    {
+        std::vector<Node<T, N>*> firstPath, secondPath;
+
+        if (!find(first, firstPath) || !find(second, secondPath))
+            return nullptr;
+
+        auto firstIt = firstPath.rbegin();
+        auto secondIt = secondPath.rbegin();
```

Comment:
[optional] The assignment asked not to store nodes in additional data structures ;)

Can you think of a way find the lowest common ancestor in constant space, without the two vectors? Hint: you might want to store a pointer to the parent in every node.